crates/turborepo-lib/src/config/override_env.rs

```diff
@@ -0,0 +1,493 @@
+        // ------------------------------------------------
+
+        // if there's no token, this is also permissible
+        if input.TURBO_TEAMID.is_some() && input.TURBO_TEAM.is_some() {
+            output.team_id = input.TURBO_TEAMID;
+            output.team_slug = input.TURBO_TEAM;
+            return output;
+        }
+
+        // handle "only" cases
+        // ------------------------------------------------
+        if input.TURBO_TEAMID.is_some() {
+            output.team_id = input.TURBO_TEAMID;
+            return output;
+        }
+
+        if input.TURBO_TEAM.is_some() {
+            output.team_slug = input.TURBO_TEAM;
+
+            if input.VERCEL_ARTIFACTS_OWNER.is_some() {
+                output.team_id = input.VERCEL_ARTIFACTS_OWNER;
+            }
+
+            return output;
+        }
+
+        if input.VERCEL_ARTIFACTS_OWNER.is_some() {
+            output.team_id = input.VERCEL_ARTIFACTS_OWNER;
+            return output;
+        }
+
+        Output::new()
+    }
+}
+
+pub struct OverrideEnvVars<'a> {
+    environment: &'a HashMap<OsString, OsString>,
+    output: Output,
+}
+
+impl<'a> OverrideEnvVars<'a> {
+    pub fn new(environment: &'a HashMap<OsString, OsString>) -> Result<Self, Error> {
+        let input = Input::from(environment);
+        let output = Output::from(input);
+
+        Ok(Self {
+            environment,
+            output,
+        })
+    }
+
+    fn ui(&self) -> Option<UIMode> {
+        // TODO double check on what's going on here
+        let value = self
+            .environment
+            .get(OsStr::new("ci"))
+            .or_else(|| self.environment.get(OsStr::new("no_color")))?;
+        match truth_env_var(value.to_str()?)? {
+            true => Some(UIMode::Stream),
+            false => None,
+        }
+    }
+}
+
+impl<'a> ResolvedConfigurationOptions for OverrideEnvVars<'a> {
+    fn get_configuration_options(
+        &self,
+        _existing_config: Option<&ConfigurationOptions>,
+    ) -> Result<ConfigurationOptions, Error> {
+        let output = ConfigurationOptions {
+            team_id: self.output.team_id.clone(),
+            token: self.output.token.clone(),
+            team_slug: self.output.team_slug.clone(),
+            ui: self.ui(),
+            api_url: None,
+            ..Default::default()
+        };
+        Ok(output)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use lazy_static::lazy_static;
+
+    use super::*;
+
+    lazy_static! {
+        static ref VERCEL_ARTIFACTS_OWNER: String = String::from("valueof:VERCEL_ARTIFACTS_OWNER");
+        static ref VERCEL_ARTIFACTS_TOKEN: String = String::from("valueof:VERCEL_ARTIFACTS_TOKEN");
+        static ref TURBO_TEAMID: String = String::from("valueof:TURBO_TEAMID");
+        static ref TURBO_TEAM: String = String::from("valueof:TURBO_TEAM");
+        static ref TURBO_TOKEN: String = String::from("valueof:TURBO_TOKEN");
+    }
```

Comment:
Not blocking, but I think this would be simplified by just using `const` and copying in the various methods
```suggestion
        const VERCEL_ARTIFACTS_OWNER: &str = "valueof:VERCEL_ARTIFACTS_OWNER";
        const VERCEL_ARTIFACTS_TOKEN: &str = "valueof:VERCEL_ARTIFACTS_TOKEN";
        const TURBO_TEAMID: &str = "valueof:TURBO_TEAMID";
        const TURBO_TEAM: &str = "valueof:TURBO_TEAM";
        const TURBO_TOKEN: &str = "valueof:TURBO_TOKEN";
```

The calls to `clone()` would be replaced by `to_owned()` or (`into()`)

Reply:
ahhhh yesssss wonderful.  I was wrestling with this for a while trying to get that to work.  Not sure, now that I see the solution, what I was stuck on, but thank you thank you 9586307e0ed2b02cd1bce4c028080052cfc48a13